exec: fallback to distsql on aggregator with no aggregate functions

yuzefovich opened the pull request:
It is possible for an aggregator to be planned without aggregate
functions, for example, in the query like:
SELECT 1 FROM t HAVING true.
It is quite tricky to make this work through vectorize, so we
should just fall back to distsql. The difficulty arises from the
fact that we need to introduce an operator that produces a batch
and takes no input (like colBatchScan) but actually does no work
(like noop) except for emitting a single batch with non-zero length.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38809)
<!-- Reviewable:end -->

asubiotto commented:
Would it be possible to easily support this? Please also expand the commit message.

yuzefovich commented:
Updated the commit message. Here it is:

It is possible for an aggregator to be planned without aggregate
functions, for example, in the query like:
SELECT 1 FROM t HAVING true.
It is quite tricky to make this work through vectorize, so we
should just fall back to distsql. The difficulty arises from the
fact that we need to introduce an operator that produces a batch
and takes no input (like colBatchScan) but actually does no work
(like noop) except for emitting a single batch with non-zero length.

I started adding support for it but was hitting null pointer errors, so I decided that it was just easier to reject such a query in vectorized. I'll give it another shot right now.

jordanlewis commented:
> The difficulty arises from the fact that we need to introduce an operator that produces a batch and takes no input (like colBatchScan) but actually does no work (like noop) except for emitting a single batch with non-zero length.

Isn't this what the various `constOp` implementations do, for the most part?

yuzefovich reviewed (commented):
`constOp`s still have an input operator, and that input decides on when to emit zero-length batch.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38809#-:-LjWr5iT38M_qqTKjvuA:b5oqib0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
The difference with `constOp` is that it is planned when we need constants to interact with other operators, for example, in query like `SELECT a + 1 FROM t`. In this case, we get a chain like `colBatchScan -> constOp -> projectionOp`.

But in this edge case (`SELECT 1 FROM t HAVING true`), `1` is in the render expressions of `Post` of aggregator spec, and we need something like `singleTupleNoInputOp -> projectionOp`. My hesitation partially comes from not knowing whether that special operator should always emit the first batch of length 1 or some other length (with the consecutive batch of length zero).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38809#-:-LjWsC8V-ptDgoyronEq:b4sc5mi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Ok, I've just pushed a second commit that actually supports this case through vectorize. Please let me know what you think.

Yesterday I ran into a few problems but didn't realize that I needed an operator that outputs a single batch with non-zero length first. I'm still unsure whether there are cases when we want the first batch to have a length greater than 1.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38809#-:-LjWwppNB5QbiKpv6Mbc:bbqbrgs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Guys, please take another look.

jordanlewis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38809)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @solongordon, and @yuzefovich)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38809#-LjwvyJBDyreLA1n01qN:-LjwvyJBDyreLA1n01qO:b-1efjhk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56c2a409d2acb312eb6de8f0eaebbb97d6b7498/pkg/sql/distsqlrun/column_exec_setup.go#L142)):*
> ```Go
> 			// SELECT 1 FROM t HAVING true. This breaks some of the assumptions, so
> 			// we'll kick the query back to DistSQL.
> 			return nil, nil, errors.Newf("aggregator with no aggregate functions is unsupported in vectorized")
> ```

Sounds good. Can we ask the optimizer team to stop planning aggregators in this case? Is it equivalent to a simple filter?


<!-- Sent from Reviewable.io -->

yuzefovich committed:
exec: handle aggregations without aggregate functions

We add a special singleTupleNoInputOperator that on the first call
to Next() outputs a batch of length 1 with no actual columns and
outputs zero-length batches on all consecutive calls. This allows
us to execute queries that have only render expressions through
vectorized engine.

Release note: None

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38809#-:-LjxLZeu1Ok9jIDjdqG-:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38809#-LjwvyJBDyreLA1n01qN:-LjwxZ8Q1kV6tFIpp0FB:b-ovpepl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56c2a409d2acb312eb6de8f0eaebbb97d6b7498/pkg/sql/distsqlrun/column_exec_setup.go#L142)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Sounds good. Can we ask the optimizer team to stop planning aggregators in this case? Is it equivalent to a simple filter?
</blockquote></details>

I pinged them in the channel. I'm not sure when this case occurs tbh.

Radu said, "`SELECT 1 FROM t HAVING true` returns 1 row regardless of how many rows there are in the table," so aggregation appears to be mandatory here.


<!-- Sent from Reviewable.io -->
